### DISCUSSION TITLE
Feedback on Microsoft.WinGet.RestSource

### DISCUSSION THREAD

**Discussion by denelon** [microsoft/winget-cli-restsource#92] [Q&A]
This discussion topic was created to provide feedback on the PowerShell module for working with REST sources.

Note, this is an initial released designed to work with the reference implementation deployed to Azure. We still have additional work to make this support any valid REST API implementation.

**Comment by jedieaston** [comment]
I clicked report but you might be able to get it moving faster:

![image](https://user-images.githubusercontent.com/21368066/140131912-71dc2f6c-b8b5-464b-bf6c-85c44ddbc9e3.png)

**Reply by denelon** [reply]
Thanks for calling this out. One of our users encountered this during some of the testing, but we were not able to reproduce.

**Reply by jedieaston** [reply]
If anyone else gets that, you have to right click on the zip, go to properties, and click "Unblock" or else the DLLs won't be loaded in PowerShell (at least, not on my system).

**Reply by doctordns** [reply]
Nice start. Three comments:
1. Please can you create a very simple installation script that enables me to just run the script. PowerShell users tend to eschew the GUI where possible. I may try to knock one up. 
2. I guess for the first set of cmdlets, I expected ones that I could use to install packages etc. These require me to use Azure. Is there a local hosting option rather than Azure  and are Azure functions for Winget Free or paid, I can't remember?
3. The Get-WInManifest cmdlet could fail more gracefully. Here is what I see:

```powershell-console
PS C:\Foo\winget\WinGet.RestSource-Winget.PowerShell.Source> Get-WinGetManifest 'FOO'
Get-AzFunctionApp_List: C:\Users\tfl.COOKHAM\Documents\PowerShell\Modules\Az.Functions\2.0.0\custom\Get-AzFunctionApp.ps1:101
Line |
 101 |  …   $apps = @(Az.Functions.internal\Get-AzFunctionApp @PSBoundParameter …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | SharedTokenCacheCredential authentication unavailable. Token acquisition failed for user
     | foo@bar.com. Ensure that you have authenticated with a developer tool that supports Azure
     | single sign on.

Exception: C:\Foo\winget\WinGet.RestSource-Winget.PowerShell.Source\Library\Get-WinGetManifest.ps1:114
Line |
 114 |  …             throw "Unable to locate Function (""$FunctionName"") in t …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Unable to locate Function ("FOO") in the Azure Tenant.
```

**Reply by jedieaston** [reply]
for 2, these aren't for interacting with winget.exe, rather they are for interacting with REST sources, so you'll need to spin up a rest source (the reference version is designed to take advantage of a bunch of Azure services, most of which have some form of local emulator that isn't supported by the scripts here).

**Reply by doctordns** [reply]
Ahh - thanks for that! As I said, I kind of expected the first cmdlets to be ones I could use in place of winget.exe.

**Reply by denelon** [reply]
I'm one PR away from opening up the new repository.

**Reply by jantari** [reply]
Is the new repository up? I didn't find it by searching at least: https://github.com/orgs/microsoft/repositories?q=winget

**Reply by doctordns** [reply]
I have not see it yet, but @denelon might have more details.

But as there is so little progress on usable PowerShell cmdlets, I wonder whether it would not just be better to cancel cmdlets totally.  Of course, that would mean giving up on the goal of having a fully functioning Packager Manager but with Chcolately, I am not sure there is really a need for yet another product (especially one that is not truly Enterprise ready).

**Reply by denelon** [reply]
We decided to keep the [cmdlets](https://github.com/microsoft/winget-cli/tree/master/tools/PowerShell) in the winget-cli repository.

**Comment by denelon** [comment]
We've been talking about local options whether it's an emulated environment, or some kind of VM or container based solution, but it's still early in the discussion pre-design phase.

**Reply by doctordns** [reply]
I am confused. I am just an IT pro that wants winget cmdlets to install stuff from a MS Curated repository. Do I really need to have an azure account and create REST data stores just to install a package?  I can use winget.exe without this stuff, so why does PowerShell cmdlets seem to need it (or why is such stuff your first priority)?

The first cmdlets should have been cmdlets that find and  install packages from the MS curated repo, as well as list, remove, or update packages that are already installed. As a first step.    

With no progress on cmdlets, I continue to be of the view that Winget is little more than a powertoy for developers and not a serious enterprise package manager platform. As ever, I am happy to be proven wrong.

**Reply by denelon** [reply]
No, you don't need an Azure account or create your own REST source. We've been working on some IT Pro based solutions. One of the areas is enterprises who wish to host their own source rather than use the Microsoft Store or the Windows Package Manager Community repository.

Building cmdlets requires a rich object model that the CLI doesn't provide. We've been refactoring the functional code into a COM API so the Windows Package Manager can simply be a client interfacing with the COM API, and PowerShell will interface with that same COM API.

**Comment by Gijsreyn** [comment]
Hi @denelon, I just stumbled on this discussion to give some small feedback on the just officially released PowerShell module `Microsoft.WinGet.RestSource`. 

When looking at both the `WinGetManifest.ps1` and `Remove-WinGetManifest.ps1` scripts, I can see that the Version or PackageVersion is a string. Perhaps it would be better to use the `[Version]` attribute to be more strongly typed. It's in the small details.

**Reply by jantari** [reply]
Unfortunately the `System.Version` type is old and extremely limited, e.g. it cannot represent semver identifiers or build metadata which are ubiquitous these days.

It would be swell if the .NET team gave us a new SemanticVersion class, but winget couldn't use it anyway because of the need to be backwards compatible with PowerShell 5.1 and because software vendors will always use weird non standard version strings that don't comply with CalVer, SemVer or anything useful.

**Reply by denelon** [reply]
The module was published as Microsoft.WinGet.RestSource at the PowerShell gallery, I'm going to update the discussion title.

**Reply by doctordns** [reply]
one great solution would be for the winget team to implement a [SemanticVersoin] class, use it here and submit it as a PR for .NET

**Reply by Gijsreyn** [reply]
@jantari, do you think the `[System.Management.Automation.SemanticVersion]` attribute would simply do the trick, as the module targets 7.4 anyway?

**Reply by jantari** [reply]
Thanks for teaching me about that, and you're right the RestSource module doesn't have to be compatible with 5.1 the way the WingetClient module has.

Imo it still wouldn't work because, again, manufacturers just do the weirdest things with their version numbers like e.g. adding underscores or using 5 numeric parts - both examples that are incompatible with both `Version` and `SemanticVersion`. Windows has allowed freeform strings for too long to take them back now.

**Reply by Gijsreyn** [reply]
Gotcha! You're right, I overlooked that part.

**Comment by Gijsreyn** [comment]
Hey @denelon, after going through the solution that is created by the `New-WinGetSource` cmdlet, I stumbled on the fact that it's using the function app host key to authenticate through APIM. Perhaps it would be an improvement to use MSI, instead of injecting a header. See screenshot below:

![image](https://github.com/user-attachments/assets/6c24a4c7-f24d-4e90-940a-5e2e7a3d3c5a)

> [!NOTE]
> It requires additional configuration to enable system or user-assigned managed identities.
